Update rustc to master

dvc94ch opened the pull request:
###### Motivation for this change

Related to #15199 

rustcBeta and rustcMaster build now. Two patches where required, one to disable the lockfile check and one to rustcMaster stage0.txt

Found two problems with buildRustPackage. Not sure exactly when and why they come up, but the changes fix them.
1. When multiple versions of the same package are required
   $revs is an array.
2. When cargo fetch is run it usually doesn't need a network
   connection. But when it does SSL_CERT_FILE isn't set.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ x Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

dvc94ch committed:
Fix buildRustPackage edge cases

1. When multiple versions of the same package are required
   $revs is an array.
2. When cargo fetch is run it usually doesn't need a network
   connection. But when it does SSL_CERT_FILE isn't set.

dvc94ch committed:
Update rustcMaster

Now builds rustcBeta (1.10.0) and rustcMaster (1.11.0).

dvc94ch committed:
Update cargo head

dvc94ch commented:
@abbradar @jgillich @the-kenny can one of you guys take a look if your not too busy? or who's boss of the rustc package? =P

dvc94ch commented:
Build failure due to lack of disk space. Should I split this into multiple smaller PRs? Or how do I get a green check mark?

the-kenny commented on pkgs/development/compilers/rustc/bootstrap.nix:
```diff
@@ -0,0 +1,49 @@
+    then "macos-x86_64"
+    else abort "no snapshot to bootstrap for this platform (missing platform url suffix)";
+
+  snapshotHashLinux686 = "0e0e4448b80d0a12b75485795244bb3857a0a7ef";
```
Are you sure that combining bootstrap for `head` and `stable` is a good idea? What if `head` doesn't build with the `stage0` for `stable` or reverse?

the-kenny replied:
Also, please don't remove the notice on how to get these hashes.

dvc94ch replied:
master and beta now build with stable, so we only need bootstrap to build stable.

the-kenny replied:
That's cool :) I didn't notice that. Can you still reintroduce the note on how to get these hashes please?

dvc94ch replied:
irrelevant since this stable is the last version that uses snapshots, so the information is outdated...

dvc94ch replied:
but I'll add it back..

the-kenny replied:
Well we have to update these hashes when stable is updated, right? Nix can't reference to our old stable version, we always have to build a bootstrap when  building stable.

dvc94ch replied:
yes, but the process of getting the hashes will be a different one...

dvc94ch replied:
and I already said I'll add it back...

the-kenny commented on pkgs/development/compilers/rustc/beta.nix:
```diff
@@ -0,0 +1,12 @@
+  srcSha = "01bx6616lslp2mbj4h8bb6m042fs0y1z8g0jgpxvbk3fbhzwafrx";
+  patches = [ ./patches/disable-lockfile-check.patch ] ++
+    stdenv.lib.optional stdenv.needsPax ./patches/grsec.patch;
+  rustc = rustcStable;
```
Wouldn't it make sense to name this `rustcStage0` or something?

dvc94ch replied:
the idea was that generic.nix depends on rustc to build rustc - and then the specific rustc that we want to build it with we override in callPackage. I don't think we need to name it rustcStage0 necessarily...

the-kenny replied:
I didn't recognize that we can pass the default `rustc` there for `beta` and `head`. Perfectly fine in this case!

the-kenny commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5260,16 +5260,18 @@ in
 
+  rustc = rustcStable;
+  rustcStable = callPackage ../development/compilers/rustc/stable.nix {};
+  rustcBeta = callPackage ../development/compilers/rustc/beta.nix {};
```
Please mark these as `lowPrio`, else `nix-env` might pick them up over `stable`

the-kenny commented:
Also, is there a reason to keep `head` in nixpkgs? The general consensus is to keep unstable packages only if there is a demand for it - especially for nightlies which have a high tendency to grow stale and even get outdated with newer releases. We don't want `rustcBeta` or `rustcHead` to be older than `rustcStable`

dvc94ch commented:
That's why I updated them, they're newer now... :) Some packages don't build without a nightly compiler. I added beta only because I thought that nightly was supposed to be built with beta, I'm not too bothered about it... The other nice thing about having a nightly is that we should be able to fix the package in advance if any changes have to be made.

the-kenny commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5260,16 +5260,18 @@ in
 
   # rust platform to build cargo itself (with cargoSnapshot)
-  rustCargoPlatform = makeRustPlatform (cargoSnapshot rustc) rustCargoPlatform;
+  rustCargoPlatform = makeRustPlatform (cargoSnapshot rustcStable) rustCargoPlatform;
```
We should go with `rustc` instead of `rustcStable` here. If users use `packageOverriders` to override `rustc`, they likely expect the other rust stuff to use the overriden `rustc` too

the-kenny commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5260,16 +5260,18 @@ in
   # rust platform to build cargo itself (with cargoSnapshot)
-  rustCargoPlatform = makeRustPlatform (cargoSnapshot rustc) rustCargoPlatform;
+  rustCargoPlatform = makeRustPlatform (cargoSnapshot rustcStable) rustCargoPlatform;
   rustUnstableCargoPlatform = makeRustPlatform (cargoSnapshot rustcMaster) rustUnstableCargoPlatform;
```
This is an opportunity to finally settle on a naming convention: `beta`, `unstable`, `head`, `master` are quite confusing. Maybe we should even remove the `rustUnstableCargoPlatform` and just keep `cargo`, `cargoUnstable` and all  `rustc`s and let users compose their own `unstable` platforms if needed.

Are there reasons not to do this?

dvc94ch replied:
I find it quite confusing too... That's why I didn't touch it =P

dvc94ch replied:
Could you please be a little more specific? I'm not 100% sure how this works and if something can be removed here...

dvc94ch replied:
what's the difference between rustPlatform and rustCargoPlatform?

dvc94ch replied:
I think the best solution would be to build cargo without cargo, I think it can be built with just make instead of a snapshot, and remove the rustCargoPlatform, but I think that may be a task for another day...

dvc94ch commented:
@the-kenny I think I took your suggestions into account and tried to improve the naming. Let me know if you have further suggestions

the-kenny commented:
Looks good to me. I'll run the builds on my system in a few minutes to see if everything works, then merge it.

dvc94ch commented:
Thanks!
On Jun 2, 2016 6:24 PM, "Moritz Ulrich" <notifications@github.com> wrote:

> Looks good to me. I'll run the builds on my system in a few minutes to see
> if everything works, then merge it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/15875#issuecomment-223344979>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAtRrzvjWf_oz6bOztp5Yr-EPqT3my3Lks5qHwO7gaJpZM4Iq2hg>
> .
>

jgillich commented:
Awesome stuff, thanks! :tada:
